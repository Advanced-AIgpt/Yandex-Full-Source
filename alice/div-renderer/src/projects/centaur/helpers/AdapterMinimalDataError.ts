export class AdapterMinimalDataError extends Error {
    constructor(missedData: string[], template: string) {
        super(`Missed data (${missedData.join(', ')}) in template ${template}`);
        this.name = 'AdapterMinimalDataError';
    }
}
