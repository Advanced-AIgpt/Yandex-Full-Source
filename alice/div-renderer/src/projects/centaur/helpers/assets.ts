export function getS3Asset(path: string) {
    return `https://dialogs.s3.yandex.net/smart_displays/${path}`;
}

export function getStaticS3Asset(path: string) {
    return `https://yastatic.net/s3/dialogs/smart_displays/${path}`;
}
