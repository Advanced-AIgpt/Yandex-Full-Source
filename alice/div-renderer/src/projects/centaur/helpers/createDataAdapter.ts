import Ajv, { AnySchema, JSONSchemaType } from 'ajv';
import { DeepNonNullable, DeepRequired } from 'ts-essentials';
import { IRequestState } from '../../../common/types/common';
import { ExpFlags } from '../expFlags';

export const ajv = new Ajv();

type DataAdapter<D, R> = (data: D, requestState: IRequestState) => R;

export const createDataAdapter = <D, R>(schema: AnySchema, dataAdapter: DataAdapter<D, R>) => {
    const validate = ajv.compile(schema);

    const resultDataAdapter: DataAdapter<D, R> = (data, requestState) => {
        const isDataValid = validate(data);

        if (isDataValid === false) {
            // TODO: CENTAUR-1243: log errors

            if (requestState.hasExperiment(ExpFlags.errorOnAssert)) {
                const [firstError] = validate.errors ?? [];

                throw new Error(`DataValidationError: "${firstError.instancePath}" ${firstError.message}`);
            }
        }

        return dataAdapter(data, requestState);
    };

    return resultDataAdapter;
};

export type MakeClear<T> = DeepRequired<DeepNonNullable<T>>;
export type MakeTestSchemaType<T> = JSONSchemaType<MakeClear<T>>;
