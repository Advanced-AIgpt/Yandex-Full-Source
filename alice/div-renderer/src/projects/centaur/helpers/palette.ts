export interface ColorSet {
    light: string;
    dark: string;
}

export interface ColorScheme {
    [k: string]: ColorSet | ColorScheme;
}

export type ColorNamespace<T extends ColorScheme> = {
    [K in keyof T]: T[K] extends ColorScheme ? ColorNamespace<T[K]> : string;
};

/** https://a.yandex-team.ru/arcadia/portal/morda-schema/api/search/2/paletted-color.json */
export interface PaletteColor {
    name: string;
    color: string;
}

export interface Palette {
    light: PaletteColor[];
    dark: PaletteColor[];
}

function isColorSet(x: ColorSet | ColorScheme): x is ColorSet {
    return Object.keys(x).length === 2 && 'light' in x && 'dark' in x;
}

function colorSchemeToPalette({
    colorScheme,
    scope,
}: {
    colorScheme: ColorScheme;
    scope: string;
}): Palette {
    const palette: Palette = {
        dark: [],
        light: [],
    };

    const stack: {
        path: string[];
        node: ColorScheme | ColorSet;
    }[] = Object.entries(colorScheme).map(([key, value]) => ({
        path: [key],
        node: value,
    }));

    while (stack.length) {
        const { path, node } = stack.pop() as {
            path: string[];
            node: ColorScheme | ColorSet;
        };

        if (isColorSet(node)) {
            const paletteName = `${scope}.${path.join('.')}`;
            palette.light.push({ name: paletteName, color: node.light });
            palette.dark.push({ name: paletteName, color: node.dark });
        } else {
            stack.push(
                ...Object.entries(node).map(([key, value]) => ({
                    path: [...path, key],
                    node: value,
                })),
            );
        }
    }

    return palette;
}

function colorSchemeToNamespace<T extends ColorScheme>({
    colorScheme,
    scope,
}: {
    colorScheme: T;
    scope: string;
}): ColorNamespace<T & typeof commonColorScheme> {
    const rootNamespace: ColorNamespace<ColorScheme> = {};

    const stack: {
        path: string[];
        key: string;
        node: ColorScheme | ColorSet;
        parentNamespace: ColorNamespace<any>;
    }[] = Object.entries(colorScheme).map(([key, value]) => ({
        key,
        path: [key],
        node: value,
        parentNamespace: rootNamespace,
    }));

    while (stack.length) {
        const { path, key, node, parentNamespace } = stack.pop() as {
            path: string[];
            key: string;
            node: ColorScheme | ColorSet;
            parentNamespace: ColorNamespace<any>;
        };

        if (isColorSet(node)) {
            parentNamespace[key] = `@{${scope}.${path.join('.')}}`;
        } else {
            const newNamespace: ColorNamespace<any> = {};
            parentNamespace[key] = newNamespace;

            stack.push(
                ...Object.entries(node).map(([key, value]) => ({
                    key,
                    path: [...path, key],
                    node: value,
                    parentNamespace: newNamespace,
                })),
            );
        }
    }

    return rootNamespace as ColorNamespace<T & typeof commonColorScheme>;
}

export function getPalette<T extends ColorScheme, U extends boolean | undefined = undefined>({
    colorScheme,
    scope,
    extendsCommon,
}: {
    colorScheme: T;
    scope: string;
    extendsCommon?: U;
}): {
    palette: Palette;
    ns: ColorNamespace<U extends true ? T & typeof commonColorScheme : T>;
} {
    if (extendsCommon) {
        colorScheme = {
            ...commonColorScheme,
            ...colorScheme,
        };
    }

    return {
        palette: colorSchemeToPalette({ colorScheme, scope }),
        ns: colorSchemeToNamespace({ colorScheme, scope }) as ColorNamespace<U extends true ? T & typeof commonColorScheme : T>,
    };
}

const commonColorScheme = {
    text: {
        primary: {
            light: '#000000',
            dark: '#ffffff',
        },
    },
};
