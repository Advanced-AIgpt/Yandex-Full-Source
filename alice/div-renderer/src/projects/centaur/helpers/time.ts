export const formatTimeUnit = (value: number): string => value >= 10 ? String(value) : '0' + value;

const minuteInSeconds = 60;
const hourInSeconds = 60 * minuteInSeconds;

export const formatTime = (value: number) => {
    const hours = Math.floor(value / hourInSeconds);
    const secondsRemainder = value % hourInSeconds;
    const minutes = Math.floor(secondsRemainder / minuteInSeconds);
    const seconds = secondsRemainder % minuteInSeconds;

    const result: string[] = [];

    if (hours > 0) {
        result.push(formatTimeUnit(hours));
    }

    result.push(formatTimeUnit(minutes));

    result.push(formatTimeUnit(seconds));

    return result.join(':');
};
