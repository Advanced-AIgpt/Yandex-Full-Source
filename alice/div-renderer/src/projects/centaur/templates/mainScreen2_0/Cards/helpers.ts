export function getCardMainScreenId({ rowIndex, colIndex }: { rowIndex?: number; colIndex?: number }) {
    return `main_screen_card_${rowIndex || 0}_${colIndex || 0}`;
}
