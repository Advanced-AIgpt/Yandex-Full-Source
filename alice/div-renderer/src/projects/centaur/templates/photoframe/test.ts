import { AnonymizeDataForSnapshot } from '../../../../common/helpers/dev';
import screenSaverRender from './screenSaver';
import { MMRequest } from '../../../../common/helpers/MMRequest';

jest.mock('../../../../common/logger');

describe('Photoframe scenario', () => {
    it('matches snapshot', () => {
        expect(AnonymizeDataForSnapshot(screenSaverRender({
            ImageUrl: 'https://dialogs.s3.yandex.net/smart_displays/photoframe/nature29.png',
        }, new MMRequest({}, {}, {})))).toMatchSnapshot();
    });
});
