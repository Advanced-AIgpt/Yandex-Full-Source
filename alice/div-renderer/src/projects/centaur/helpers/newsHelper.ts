import { getStaticS3Asset } from './assets';

const TOPICS = new Set([
    'auto',
    'business',
    'computers',
    'culture',
    'incident',
    'index',
    'koronavirus',
    'personal',
    'politics',
    'science',
    'society',
    'sport',
    'world',
]);

const ALIASES = {
    nplus1: 'science',
} as { [name: string]: string };

export function getImageByTopic(topic: string | null | undefined): string {
    let resultImage = 'news/index.png';

    if (topic) {
        if (Object.prototype.hasOwnProperty.call(ALIASES, topic)) {
            topic = ALIASES[topic];
        }

        if (TOPICS.has(topic)) {
            resultImage = `news/${topic}.png`;
        }
    }
    return getStaticS3Asset(resultImage);
}
