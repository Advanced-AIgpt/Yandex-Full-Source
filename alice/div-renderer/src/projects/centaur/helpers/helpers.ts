/* eslint-disable camelcase */
import { sortedIndexBy } from 'lodash';
import { format, isSameDay } from 'date-fns';
import { ru } from 'date-fns/locale';
import { Div, IDivData, TextBlock } from 'divcard2';
import { IRequestState } from '../../../common/types/common';
import { getTemplates, templatesRegistry } from '../../../registries/common';
import { centaurTemplatesClass } from '../index';
import { Palette } from './palette';

export { default as pluralize } from '@yandex-int/i18n/lib/plural/ru';
type TextProps = Pick<ConstructorParameters<typeof TextBlock>[0], 'text_color'|'font_size'|'font_weight'|'line_height'>;

export const textBreakpoint =
    (...breakpoints: [number, TextProps][]) =>
        (text: string) => {
            const idx = Math.min(sortedIndexBy(breakpoints, [text.length, {}], 0), breakpoints.length - 1);
            return { text, ...breakpoints[idx][1] };
        };

export const formatDate = (d: Date | string | number, f: string) => format(new Date(d), f, { locale: ru });

export const textCommand = (c: string) => `dialog://text_command?query=${encodeURI(c)}`;

export const updatedAt = (d: Date) => {
    const now = new Date();
    return `Обновлен ${isSameDay(now, d) ? 'сегодня' : formatDate(d, 'dd.MM')}`;
};

export const insertBetween = <I extends unknown, S extends unknown>(itemList: I[], separator: S) => {
    return itemList.reduce<(I | S)[]>((result, item, index) => {
        if (index !== 0) {
            result.push(separator);
        }

        result.push(item);

        return result;
    }, []);
};

export interface ITemplateCard {
    templates: {
        [name: string]: Div;
    };
    palette?: Palette;
    card: IDivData;
}

// Дива верхнего уровня. Замена для new `TemplateCard()`
export const TopLevelCard = (card: Omit<IDivData, 'variable_triggers' | 'variables'>, requestState: IRequestState, palette?: Palette) => {
    const templates = Object.assign(
        getTemplates(requestState.localTemplates, templatesRegistry),
        centaurTemplatesClass.getUsedTemplates(requestState),
        requestState.res.localTemplates.getAll(),
    );

    return {
        templates,
        palette,
        card: {
            ...card,
            variable_triggers: requestState.variableTriggers.getAll(),
            variables: [...requestState.variables],
        } as IDivData,
    };
};
