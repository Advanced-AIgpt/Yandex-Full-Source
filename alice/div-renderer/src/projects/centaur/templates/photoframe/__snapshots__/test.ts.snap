// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Photoframe scenario matches snapshot 1`] = `
Object {
  "card": Object {
    "log_id": "gallery-screensaver",
    "states": Array [
      Object {
        "div": Object {
          "background": Array [
            Object {
              "image_url": "https://dialogs.s3.yandex.net/smart_displays/photoframe/nature29.png",
              "preload_required": 1,
              "type": "image",
            },
            Object {
              "color": "#2E000000",
              "type": "solid",
            },
            Object {
              "angle": 74,
              "colors": Array [
                "#29000000",
                "#1F000000",
                "#00000000",
              ],
              "type": "gradient",
            },
          ],
          "delimiter_style": Object {
            "color": "#0fff",
          },
          "height": Object {
            "type": "match_parent",
          },
          "type": "separator",
          "width": Object {
            "type": "match_parent",
          },
        },
        "state_id": 0,
      },
    ],
  },
  "templates": Object {},
}
`;
